docs/sender.rst

```diff
@@ -0,0 +1,409 @@
+.. _sender:
+
+=====================
+Sending Data over ILP
+=====================
+
+The :class:`questdb.ingress.Sender` class is a client that inserts rows into
+QuestDB via the ILP protocol.
+
+It supports both TCP and HTTP protocols, authentication and TLS.
+
+You should prefer to use HTTP over TCP in most cases as it provides better
+feedback on errors and transaction control.
+
+TCP is useful for high-throughput scenarios in higher latency networks.
+
+Basic Usage
+===========
+
+In this example, data will be flushed and sent at the end of the ``with``
+block.
+
+.. code-block:: python
+
+    from questdb.ingress import Sender, TimestampNanos
+
+    conf = 'http::addr=localhost:9000;'
+    with Sender.from_conf(conf) as sender:
+        sender.row(
+            'weather_sensor',
+            symbols={'id': 'toronto1'},
+            columns={'temperature': 23.5, 'humidity': 0.49},
+            at=TimestampNanos.now())
+        sensor.row(
+            'weather_sensor',
+            symbols={'id': 'dubai2'},
+            columns={'temperature': 41.2, 'humidity': 0.34},
+            at=TimestampNanos.now())
+
+The ``Sender`` object holds an internal buffer. The call to ``.row()``
+simply forwards all arguments to the :func:`Buffer.row` method.
+
+The ``Sender`` class is generally initialised with a configuration string.
+
+The format of the configuration string is::
+
+    <protocol>::<key>=<value>;<key>=<value>;...;
+
+The valid protocols are:
+
+* ``tcp``: ILP/TCP
+* ``tcps``: ILP/TCP with TLS
+* ``http``: ILP/HTTP
+* ``https``: ILP/HTTP with TLS
+
+Only the ``addr=host:port`` key is mandatory. It specifies the hostname and port
+of the QuestDB server. If omitted, the port will be defaulted to 9009 for TCP(s)
+and 9000 for HTTP(s).
+
+The full set of possible ``key=value;`` settings is documented in the
+:ref:`configuring-sender` section.
+
+Preparing Data
+==============
+
+Sending Rows
+------------
+
+You can send as many rows as you like by calling the
+:func:`questdb.ingress.Sender.row` method. The full method arguments are
+documented in the :func:`questdb.ingress.Buffer.row` method.
+
+Sending Pandas Dataframes
+-------------------------
+
+The sender can also send data from a Pandas dataframe.
+
+.. literalinclude:: ../examples/pandas_basic.py
+   :language: python
+
+For more details see :func:`questdb.ingress.Sender.dataframe`
+and for full argument options see :func:`questdb.ingress.Buffer.dataframe`.
+
+Populating Timestamps
+---------------------
+
+The ``at`` parameter is used to specify the timestamp of the row. It can be
+either a ``TimestampNanos`` object or a ``datetime.datetime`` object.
```

Comment:
Because people will mess up and get stuff wrong. What if they used `time.time_ns()` instead?
Instead users are forced, by design, to pick a scale - this is not new.